### ISSUE TITLE
symm tactic can incorrectly introduce a recursive hypothesis

### ISSUE THREAD

**Issue by dwrensha** [leanprover-community/mathlib4#1061]
Consider the following Lean code:
```lean
import Mathlib.Logic.Basic
import Mathlib.Tactic.Relation.Symm

variable {p : Prop}
variable (h : p ↔ p)

theorem foo : p :=
by symm at h -- (incorrectly) introduces a recursive hypothesis `foo : p`.
   assumption -- no error here

/-
all goals are closed, and we get an error:

  fail to show termination for
    foo
  with errors
  structural recursion cannot be used

  well-founded recursion cannot be used, 'foo' does not take any (non-fixed) arguments
-/
```

I expect this to give me an error at `assumption`. Instead, we get the above-shown error about structural recursion.

**Comment by dwrensha** [comment]
I believe the bad recursive hypothesis is being introduced in the `replace` call here: https://github.com/leanprover-community/mathlib4/blob/1305c7997e8133859d08c576d278bdfb1bfc30b6/Mathlib/Tactic/Relation/Symm.lean#L67

**Comment by dwrensha** [comment]
I suspect the problem has to do with [`assertAfter`](https://github.com/leanprover/lean4/blob/f6cd6c069587cfe62dd68cb6330f9ad794a56724/src/Lean/Meta/Tactic/Assert.lean#L70-L95) not handling auxDecls correctly.

**Comment by dwrensha** [comment]
(This came up in my port of `tauto`, which like `symm` uses the [`replace`](https://github.com/leanprover-community/mathlib4/blob/39fe9f89b0d0c6519caf6a84642545af3d63a01b/Mathlib/Lean/Meta.lean#L17-L42) function.)

**Comment by dwrensha** [comment]
Here's a more self-contained example:

```lean
import Lean
import Mathlib.Lean.Meta

open Lean Elab.Tactic Parser.Tactic Lean.Meta

syntax (name := testReplace) "test_replace " haveDecl : tactic

elab_rules : tactic
  | `(tactic| test_replace $n:ident := $v:term) => withMainContext do
  let name : Name := n.getId
  let e ← elabTerm v none
  let ty ← inferType e
  let hId? := (← getLCtx).findFromUserName? name |>.map fun d ↦ d.fvarId
  match hId? with
  | some hId => do
     liftMetaTactic fun mvar => do
       let result ← mvar.replace hId ty e
       pure [result.mvarId]
     pure ()
  | none     => pure ()

variable {p : Prop}
variable (h : p ↔ p)

theorem foo : p :=
by test_replace h := h
   sorry

/-
  at the `sorry`, the context is:

    Goals (1)
    p : Prop
    h : p ↔ p
    foo : p
    ⊢ p

That `foo` hypothesis should not be there.
-/

```

**Comment by dwrensha** [comment]
Reported upstream as https://github.com/leanprover/lean4/issues/1963

**Comment by dwrensha** [comment]
This is fixed now, presumably by #1157, which pulls in https://github.com/leanprover/lean4/pull/1968. Thanks @gebner!
